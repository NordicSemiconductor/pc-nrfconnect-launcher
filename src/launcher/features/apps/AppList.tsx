/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';

import { App as AppType, isInstalled } from '../../../ipc/apps';
import { useLauncherSelector } from '../../util/hooks';
import WithScrollbarContainer from '../../util/WithScrollbarContainer';
import AppFilterBar from '../filter/AppFilterBar';
import { getAppsFilter } from '../filter/filterSlice';
import ReleaseNotesDialog from '../releaseNotes/ReleaseNotesDialog';
import App from './App/App';
import { getAllApps } from './appsSlice';
import ConfirmLaunchDialog from './ConfirmLaunchDialog';
import InstallOtherVersionDialog from './InstallOtherVersionDialog';

const sortByStateAndName = (appA: AppType, appB: AppType) => {
    const cmpInstalled = Number(isInstalled(appB)) - Number(isInstalled(appA));

    const aName = appA.displayName || appA.name;
    const bName = appB.displayName || appB.name;

    return cmpInstalled || aName.localeCompare(bName);
};

export default () => {
    const allApps = useLauncherSelector(getAllApps);
    const appsFilter = useLauncherSelector(getAppsFilter);

    const apps = allApps.filter(appsFilter).sort(sortByStateAndName);

    return (
        <>
            <AppFilterBar />
            <WithScrollbarContainer hasFilter>
                {apps.map(app => (
                    <App key={`${app.name}-${app.source}`} app={app} />
                ))}
            </WithScrollbarContainer>

            <ConfirmLaunchDialog />
            <InstallOtherVersionDialog />
            <ReleaseNotesDialog />
        </>
    );
};
