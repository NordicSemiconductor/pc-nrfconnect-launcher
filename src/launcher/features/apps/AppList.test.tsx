/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';

import {
    createDownloadableTestApp,
    createLocalTestApp,
    createUninstalledTestApp,
} from '../../../test/testFixtures';
import render, { preparedStore } from '../../../test/testrenderer';
import AppList from './AppList';
import {
    checkEngineAndLaunch,
    installDownloadableApp,
    removeDownloadableApp,
} from './appsEffects';
import {
    addDownloadableApps,
    installDownloadableAppStarted,
    removeDownloadableAppStarted,
    setAllLocalApps,
    updateDownloadableAppStarted,
} from './appsSlice';

// Do not render react-bootstrap components in tests
jest.mock('react-bootstrap', () => ({
    Modal: 'Modal',
    Button: 'Button',
    ModalHeader: 'ModalHeader',
    ModalFooter: 'ModalFooter',
    ModalBody: 'ModalBody',
    ModalTitle: 'ModalTitle',
}));

jest.mock('../../features/filter/AppFilterBar', () => 'div');
jest.mock('../../features/releaseNotes/ReleaseNotesDialog', () => 'div');
jest.mock('../../util/mainConfig', () => () => ({ version: '6.1.0' }));

jest.mock('../../features/apps/appsEffects');

const localApp = createLocalTestApp('dummy', { engineVersion: '6.1.0' });

const updatableApp = createDownloadableTestApp('updatable', {
    currentVersion: '1.2.3',
    latestVersion: '1.2.4',
    engineVersion: '6.1.0',
});

const uninstalledApp = createUninstalledTestApp('uninstalled');

const installedApp = createDownloadableTestApp('installed', {
    engineVersion: '6.1.0',
});

const mockThunk = (thunkToMock: jest.MockableFunction) => {
    jest.mocked(thunkToMock).mockReturnValue({
        type: 'an action',
    });
};

describe('AppList', () => {
    it('should render without any apps', () => {
        expect(render(<AppList />).baseElement).toMatchSnapshot();
    });

    it('should render local, not-installed, installed, and updatable apps', () => {
        expect(
            render(<AppList />, [
                addDownloadableApps([
                    uninstalledApp,
                    installedApp,
                    updatableApp,
                ]),
                setAllLocalApps([localApp]),
            ]).baseElement
        ).toMatchSnapshot();
    });

    it('should disable buttons and display "Installing..." button text while installing an app', () => {
        expect(
            render(<AppList />, [
                addDownloadableApps([uninstalledApp]),
                installDownloadableAppStarted(uninstalledApp),
            ]).baseElement
        ).toMatchSnapshot();
    });

    it('should disable buttons while removing an app', () => {
        expect(
            render(<AppList />, [
                addDownloadableApps([installedApp]),
                removeDownloadableAppStarted(installedApp),
            ]).baseElement
        ).toMatchSnapshot();
    });

    it('should disable buttons and display "Updating..." while updating an app', () => {
        expect(
            render(<AppList />, [
                addDownloadableApps([updatableApp]),
                updateDownloadableAppStarted(updatableApp),
            ]).baseElement
        ).toMatchSnapshot();
    });

    it('should invoke installDownloadableApp with app name and source when install button is clicked', () => {
        mockThunk(installDownloadableApp);

        const wrapper = mount(
            <Provider
                store={preparedStore([addDownloadableApps([uninstalledApp])])}
            >
                <AppList />
            </Provider>
        );
        wrapper
            .find(`[title="Install ${uninstalledApp.displayName}"]`)
            .first()
            .simulate('click');

        expect(installDownloadableApp).toHaveBeenCalledWith(
            expect.objectContaining(uninstalledApp)
        );
    });

    it('should invoke removeDownloadableApp with app name and source when remove button is clicked', () => {
        mockThunk(removeDownloadableApp);

        const wrapper = mount(
            <Provider
                store={preparedStore([addDownloadableApps([installedApp])])}
            >
                <AppList />
            </Provider>
        );

        wrapper.find(`button.dropdown-toggle`).first().simulate('click');
        wrapper
            .find(`[title="Remove ${installedApp.displayName}"]`)
            .first()
            .simulate('click');

        expect(removeDownloadableApp).toHaveBeenCalledWith(
            expect.objectContaining(installedApp)
        );
    });

    it('should invoke checkEngineAndLaunch with given app item when Open is clicked', () => {
        mockThunk(checkEngineAndLaunch);

        const wrapper = mount(
            <Provider
                store={preparedStore([addDownloadableApps([installedApp])])}
            >
                <AppList />
            </Provider>
        );

        wrapper
            .find(`[title="Open ${installedApp.displayName}"]`)
            .first()
            .simulate('click');

        expect(checkEngineAndLaunch).toHaveBeenCalledWith(
            expect.objectContaining(installedApp)
        );
    });
});
