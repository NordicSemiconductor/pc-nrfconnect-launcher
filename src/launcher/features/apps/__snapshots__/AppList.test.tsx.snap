// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppList should disable buttons and display "Installing..." button text while installing an app 1`] = `
<body>
  <div>
    <div />
    <div
      class="with-scrollbar filter-adjusted-height"
    >
      <div
        class="content-container"
      >
        <div
          class="list-group-item"
        >
          <div
            class="py-1 row no-gutters"
          >
            <div
              class="d-flex align-items-start my-2 mr-3 col-auto"
            >
              <div
                class="core-app-icon"
              >
                <div
                  class="icon-replacement"
                />
              </div>
            </div>
            <div
              class="col"
            >
              <div
                class="h8"
              >
                App B
              </div>
              <div
                class="small text-muted"
              >
                appB description
              </div>
              <div
                class="small text-muted-more"
              >
                official
                
              </div>
            </div>
            <div
              class="d-flex align-items-center my-3 pl-3 ml-auto col-auto"
            >
              <div
                class="wide-btns btn-toolbar"
                role="toolbar"
              >
                <button
                  class="btn btn-outline-secondary"
                  disabled=""
                  title="Install App B"
                  type="button"
                >
                  Installing...
                </button>
                <div
                  class="dropdown"
                >
                  <button
                    aria-expanded="false"
                    aria-haspopup="true"
                    class="dropdown-toggle btn btn-outline-secondary"
                    type="button"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div />
  </div>
</body>
`;

exports[`AppList should disable buttons and display "Updating..." while updating an app 1`] = `
<body>
  <div>
    <div />
    <div
      class="with-scrollbar filter-adjusted-height"
    >
      <div
        class="content-container"
      >
        <div
          class="list-group-item"
        >
          <div
            class="py-1 row no-gutters"
          >
            <div
              class="d-flex align-items-start my-2 mr-3 col-auto"
            >
              <div
                class="core-app-icon"
              >
                <div
                  class="icon-replacement"
                />
              </div>
            </div>
            <div
              class="col"
            >
              <div
                class="h8"
              >
                App A
              </div>
              <div
                class="small text-muted"
              >
                appA description
              </div>
              <div
                class="small text-muted-more"
              >
                official
                , v
                1.2.3
                 (v
                1.2.4
                 available)
              </div>
            </div>
            <div
              class="d-flex align-items-center my-3 pl-3 ml-auto col-auto"
            >
              <div
                class="wide-btns btn-toolbar"
                role="toolbar"
              >
                <button
                  class="btn btn-outline-primary"
                  disabled=""
                  title="Update App A"
                  type="button"
                >
                  Updating...
                </button>
                <button
                  class="btn btn-primary"
                  disabled=""
                  title="Open App A"
                  type="button"
                >
                  Open
                </button>
                <div
                  class="dropdown"
                >
                  <button
                    aria-expanded="false"
                    aria-haspopup="true"
                    class="dropdown-toggle btn btn-outline-primary"
                    type="button"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div />
  </div>
</body>
`;

exports[`AppList should disable buttons while removing an app 1`] = `
<body>
  <div>
    <div />
    <div
      class="with-scrollbar filter-adjusted-height"
    >
      <div
        class="content-container"
      >
        <div
          class="list-group-item"
        >
          <div
            class="py-1 row no-gutters"
          >
            <div
              class="d-flex align-items-start my-2 mr-3 col-auto"
            >
              <div
                class="core-app-icon"
              >
                <div
                  class="icon-replacement"
                />
              </div>
            </div>
            <div
              class="col"
            >
              <div
                class="h8"
              >
                App C
              </div>
              <div
                class="small text-muted"
              >
                appC description
              </div>
              <div
                class="small text-muted-more"
              >
                official
                , v
                1.2.3
              </div>
            </div>
            <div
              class="d-flex align-items-center my-3 pl-3 ml-auto col-auto"
            >
              <div
                class="wide-btns btn-toolbar"
                role="toolbar"
              >
                <button
                  class="btn btn-primary"
                  disabled=""
                  title="Open App C"
                  type="button"
                >
                  Open
                </button>
                <div
                  class="dropdown"
                >
                  <button
                    aria-expanded="false"
                    aria-haspopup="true"
                    class="dropdown-toggle btn btn-outline-primary"
                    type="button"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div />
  </div>
</body>
`;

exports[`AppList should render local, not-installed, installed, and upgradable apps 1`] = `
<body>
  <div>
    <div />
    <div
      class="with-scrollbar filter-adjusted-height"
    >
      <div
        class="content-container"
      >
        <div
          class="list-group-item"
        >
          <div
            class="py-1 row no-gutters"
          >
            <div
              class="d-flex align-items-start my-2 mr-3 col-auto"
            >
              <div
                class="core-app-icon"
              >
                <div
                  class="icon-replacement"
                />
              </div>
            </div>
            <div
              class="col"
            >
              <div
                class="h8"
              >
                App A
              </div>
              <div
                class="small text-muted"
              >
                appA description
              </div>
              <div
                class="small text-muted-more"
              >
                official
                , v
                1.2.3
                 (v
                1.2.4
                 available)
              </div>
            </div>
            <div
              class="d-flex align-items-center my-3 pl-3 ml-auto col-auto"
            >
              <div
                class="wide-btns btn-toolbar"
                role="toolbar"
              >
                <button
                  class="btn btn-outline-primary"
                  title="Update App A"
                  type="button"
                >
                  Update
                </button>
                <button
                  class="btn btn-primary"
                  title="Open App A"
                  type="button"
                >
                  Open
                </button>
                <div
                  class="dropdown"
                >
                  <button
                    aria-expanded="false"
                    aria-haspopup="true"
                    class="dropdown-toggle btn btn-outline-primary"
                    type="button"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="list-group-item"
        >
          <div
            class="py-1 row no-gutters"
          >
            <div
              class="d-flex align-items-start my-2 mr-3 col-auto"
            >
              <div
                class="core-app-icon"
              >
                <div
                  class="icon-replacement"
                />
              </div>
            </div>
            <div
              class="col"
            >
              <div
                class="h8"
              >
                App C
              </div>
              <div
                class="small text-muted"
              >
                appC description
              </div>
              <div
                class="small text-muted-more"
              >
                official
                , v
                1.2.3
              </div>
            </div>
            <div
              class="d-flex align-items-center my-3 pl-3 ml-auto col-auto"
            >
              <div
                class="wide-btns btn-toolbar"
                role="toolbar"
              >
                <button
                  class="btn btn-primary"
                  title="Open App C"
                  type="button"
                >
                  Open
                </button>
                <div
                  class="dropdown"
                >
                  <button
                    aria-expanded="false"
                    aria-haspopup="true"
                    class="dropdown-toggle btn btn-outline-primary"
                    type="button"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="list-group-item"
        >
          <div
            class="py-1 row no-gutters"
          >
            <div
              class="d-flex align-items-start my-2 mr-3 col-auto"
            >
              <div
                class="core-app-icon"
              >
                <div
                  class="icon-replacement"
                />
              </div>
            </div>
            <div
              class="col"
            >
              <div
                class="h8"
              >
                Local App
              </div>
              <div
                class="small text-muted"
              >
                An local app
              </div>
              <div
                class="small text-muted-more"
              >
                local
                , v
                2.0.0
              </div>
            </div>
            <div
              class="d-flex align-items-center my-3 pl-3 ml-auto col-auto"
            >
              <div
                class="wide-btns btn-toolbar"
                role="toolbar"
              >
                <button
                  class="btn btn-primary"
                  title="Open Local App"
                  type="button"
                >
                  Open
                </button>
                <div
                  class="dropdown"
                >
                  <button
                    aria-expanded="false"
                    aria-haspopup="true"
                    class="dropdown-toggle btn btn-outline-primary"
                    type="button"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="list-group-item"
        >
          <div
            class="py-1 row no-gutters"
          >
            <div
              class="d-flex align-items-start my-2 mr-3 col-auto"
            >
              <div
                class="core-app-icon"
              >
                <div
                  class="icon-replacement"
                />
              </div>
            </div>
            <div
              class="col"
            >
              <div
                class="h8"
              >
                App B
              </div>
              <div
                class="small text-muted"
              >
                appB description
              </div>
              <div
                class="small text-muted-more"
              >
                official
                
              </div>
            </div>
            <div
              class="d-flex align-items-center my-3 pl-3 ml-auto col-auto"
            >
              <div
                class="wide-btns btn-toolbar"
                role="toolbar"
              >
                <button
                  class="btn btn-outline-secondary"
                  title="Install App B"
                  type="button"
                >
                  Install
                </button>
                <div
                  class="dropdown"
                >
                  <button
                    aria-expanded="false"
                    aria-haspopup="true"
                    class="dropdown-toggle btn btn-outline-secondary"
                    type="button"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div />
  </div>
</body>
`;

exports[`AppList should render without any apps 1`] = `
<body>
  <div>
    <div />
    <div
      class="with-scrollbar filter-adjusted-height"
    >
      <div
        class="content-container"
      />
    </div>
    <div />
  </div>
</body>
`;
